### DISCUSSION TITLE
🪵 Logging518 Road to v1.0.0

### DISCUSSION THREAD

**Discussion by mharrisb1** [mharrisb1/logging518#11] [Ideas]
# 🪵 Logging518 Enhancements and the Road to v1.0.0

@smithk86 @Nagidal

Thanks for taking the time to suggest enhancements for this project. This project was basically abandoned because it worked for my specific needs at a specific point in time. 

However, given that both of you saw some value in the project and the obvious need to allow Python’s native logging module to be configured in a unified pyproject.toml config file, I would like to make it more usable for you and other members of the Python community.

## Building a good foundation

I currently work full-time and cannot promise more than 5 hours/week to this project. I would like to enable people to make contributions via a clear, well-understood process that involves very little of my time to review and approve.

To do that, I need to create a solid foundation on which people can build. What does that foundation entail?

### 1. Statement of Purpose/Letter of Intent

There needs to be a clear and concise statement at the top of the README that says why this project exists, what it does, and what it doesn’t/won’t do.

The last part is very important. I have many common logging utility modules and functions living in my past projects. Things to help with structured log messages for parsing, deriving runtime stats for observability, call lineage for traceability, etc.

I generally have versions (or exact copies) of these utilities in all of my projects and they are obviously very useful. But, what I don’t want to happen is for this project to become an overloaded collection of helpful logging utilities.

### 2. Code Conventions

Enforcing the use of Black and MyPy with CI checks and pre-commits and having clearly stated rules on function and variable names will ensure that discussions around contributions don’t focus what the code looks like, but what the code does.

### 3. Coverage

Enforcing docs and code coverage will ensure that the contributors intent (if not already obvious from the main code) is immediately understood. This will also help with change management.

### 4. Release Process

Using a standard semantic release process is a given but also including a CHANGELOG and release notes should start happening.

Also I’d like to start thinking about branching strategies. Currently, my plan is to protect `main` and any `release/*` branch and allow PRs directly to `main` without the need for `qa`, `dev`, `nightly`, or any other intermediate branching. But, I am open to feedback on this.

## Where we’re at vs. where we’re going

Currently this library was built for simple projects that made use of the [tool table](https://peps.python.org/pep-0518/#tool-table) in a pyproject.toml at the root of the application. It’s built for this use case because this was my specific use case at the time. The user experience was also built around what I felt was sufficient for myself at the time.

I want to expand this to be more flexible so it can fit other developer’s use cases as well. To do that I need to actually hear more use cases and do some research on how larger projects, especially developer libraries, use and configure logging.

@smithk86 I like your callback idea and it’s unfortunate that you had to create a new project to get this to work for your use case. I’d like to incorporate your idea in to this project. To do that, I’d like to hear more about what you’re trying to do and why you can’t do that today. Especially if you’re not able to do that with the native `logging` module through legacy configuration options. Let’s use this discussion to talk about your idea.

@Nagidal I also like your idea. The vision of this is to have zero third-party dependencies so the fact that we have to include `toml` is not ideal even though the library is small. I am looking in to `tomli` and [PEP 680](https://peps.python.org/pep-0680/) so I can add that in.

Please let me know what you’d like to see from this project.

Right now, the sole intention is to have this be a simple, light-weight utility for deserializing a TOML file into a Python Dict that can then be passed to the logging modules `dictConfig()`.

Is that the right approach or would you rather it become a further abstraction for logging similar to [loguru](https://github.com/Delgan/loguru)?

**Comment by mharrisb1** [comment]
## 💡 Idea

Remove wrapper around `Logger` class and only wrap around `logging.config`.

### Usage

As a user, I would use `logging` the way I normally would but instead of importing `logging.config`, I would import `logging518.config`.

From this:

```python
import logging
import logging.config

logging.config.fileConfig("logging.conf")
logger = logging.getLogger("project")
```

To this:

```python
import logging
import logging518.config

logging518.config.fileConfig("pyproject.toml")
logger = logging.getLogger("project")
```

**Reply by smithk86** [reply]
I think this is absolutely the right direction as it mimics the other configuration methods in `logging.config`.

**Comment by smithk86** [comment]
I believe most projects benefit from being as simple as possible. For `logging518`, I think getting logging configurations from the project's central config file (`pyproject.toml` or similarly named toml file) to `dictConfig()` should be its sole job.

I do not see any reason for it to have any more features. Additional modules could always pull in loggin518 and include more advanced utilities. I also love the idea of using tomllib for Python 3.11 and beyond.

My use case 90% of the time will be configuring logging on lightweight web applications. Flask and Quart both do some bootstrapping of the logging system but do not configure a method to output non-Flask/non-Quart logs to a file; I find their built-in log configurations helpful but insufficient. At runtime, I pull values from `os.environ` so the `FileHandler` file name can be modified for different environments.

The code I wrote for [logging518ng](https://github.com/smithk86/logging518ng/) should do a pretty good job of summarizing what I need from this project:

1. Read and load logging configurations from `pyproject.toml`
2. Control when `dictConfig()` is called. Having it run when the module is imported is probably fine for a lot of projects; whether the configs are loaded implicitly (at import) or explicitly could be controlled via the toml file.
4. A method of modifying the configuration before it is loaded. This is important as there are a number of logging configuration, like file paths, that cannot be changed after the handlers are created.
5. _optional_: Seemless integration with the built-in `logging` module so the user only needs to import `loggin518`. This is a helpful feature but not a requirement.

I have a project called "quart-template" I created to show some co-workers how I setup Quart apps. I have included `logging518ng` in this project: https://github.com/smithk86/quart-template/blob/main/app/__init__.py

Thanks for reaching out! I started my own project out of necessity but I would much rather get this one to a point that my code is unnecessary. I also have no plans to push `logging518ng` to PyPI.

Cheers!
Kyle

**Reply by mharrisb1** [reply]
Thanks @smithk86! 

I think the current plan then will be to first mimic the current `fileConfig` [implementation](https://github.com/python/cpython/blob/9a458934f7be37c59525350f1f9ecbdcd69903c1/Lib/logging/config.py#L51) in the standard lib but have it parse TOML files. That would remove the use of `dictConfig` in Logging518.

For that I'd also bring in `tomlib` instead of `toml` since that will make it in to the standard lib in 3.11.

I think once `tomlib` is in the standard library there should be no reason that what Logging518 is trying to accomplish isn't added to the actual `fileConfig` which would hopefully make this project redundant (which is oddly the goal).

Looking at your requirements above with that initial plan in mind:

1. Read and load logging configurations from `pyproject.toml` ✅ 
2. Control when `dictConfig()` is called 🤷‍♂️ 
3. A method of modifying the configuration before it is loaded 🤷‍♂️ 
4. Seemless integration with the built-in logging module ✅  

For 2, would using the example interface proposal above (where the user just replaces `logging.config`) be sufficient? The user would have to explicitly call `logging518.config.fileConfig()` to load the config file

For 3, I see how you're using this in your example project that you shared. Just so I'm on the same page, is the flow: lazy load the config, parse and potentially update the config, and then actually load the config?

For 4, I could essentially make `logging518` a mirror of `logging` where all I do is override the `fileConfig` function. That should allow the user to do everything they can do with the native logging module.

I know 4 is a nice-to-have rather than essential but I do see your point from a UX perspective.

Instead of:

```python
import logging
import logging518.config

logging518.config.fileConfig('pyproject.toml')
logger = logging.getLogger('project')
```

The user would do:

```python
import logging518
import logging518.config

logging518.config.fileConfig('pyproject.toml')
logger = logging518.getLogger('project')
```

which is slightly more preferable since it is uniform.

**Comment by smithk86** [comment]
> For 2, would using the example interface proposal above (where the user just replaces logging.config) be sufficient? The user would have to explicitly call logging518.config.fileConfig() to load the config file

I think this sounds great. I like following the pattern of the `logging.config`

> For 3, I see how you're using this in your example project that you shared. Just so I'm on the same page, is the flow: lazy load the config, parse and potentially update the config, and then actually load the config?

Yes. You have this correct. I really am doubting though if this should be a feature. It is a good feature for me but it very well may be too much of an edge case. `logging.config.fileConfig` does not handle this kind of functionality. I may try creating a patch for logging.config.dictConfig using something like [gorilla](https://pypi.org/project/gorilla/) to intercept the configs before they are loaded.

> For 4, I could essentially make logging518 a mirror of logging where all I do is override the fileConfig function. That should allow the user to do everything they can do with the native logging module.

I think trying to mirror `logging` could cause confusion. If the loading function is moved to `logging518.config`, it seems best to just focus on loading the toml and use `logging` for everything else.
